aten/src/ATen/native/quantized/QTensor.cpp

```diff
@@ -232,5 +232,93 @@ std::tuple<double, int64_t> _choose_qparams_per_tensor(
   return std::make_tuple(q_params.scale, q_params.zero_point);
 }
 
+float calculate_quant_loss(
+    const float* input,
+    int numel,
+    float xmin,
+    float xmax,
+    float* q_input,
+    int bit_rate) {
+  xmin = static_cast<at::Half>(xmin);
+  float data_range = xmax - xmin;
+  float qmax = (1 << bit_rate) - 1;
+  float scale = data_range == 0
+      ? 1.0
+      : static_cast<float>(static_cast<at::Half>(data_range / qmax));
+  float inverse_scale = 1.0f / scale;
+
+  float norm = 0.0f;
+  constexpr int VLEN = 8;
+  int i = 0;
+
+// TODO add FBGEMM kernel
+// #ifdef USE_FBGEMM
+// #endif
+
+  // remainder loop
+  for (; i < numel; i++) {
+    q_input[i] = std::max(
+        0.0f, std::min<float>(nearbyint((input[i] - xmin) * inverse_scale), qmax));
+    q_input[i] = q_input[i] * scale + xmin;
+    norm += (input[i] - q_input[i]) * (input[i] - q_input[i]);
+  }
+  return std::sqrt(norm);
+}
+
+/*
+  Helper function to find the best min/max for a tensor to calculate qparams.
+  It uses a greedy approach to nudge the min and max and calculate the l2 norm
+  and tries to minimize the quant error by doing `torch.norm(x-fake_quant(x,s,z))`
+  Returns the optimized xmax and xmin value of the tensor.
+*/
+std::tuple<double, double> choose_qparams_optimized(
+    const at::Tensor& input_tensor,
+    int64_t numel,
+    const int64_t n_bins,
+    const double ratio,
+    double xmin,
+    double xmax,
+    int64_t bit_rate) {
+
+  xmin = (float) xmin;
```

Comment:
Do we need xmin and xmax as inputs? Can we compute them from the input tensor to start the optimization?